### PR TITLE
[Doc Rework] - implementation notes PR1

### PR DESCRIPTION
**Pull request by mapedraza**
This PR is part of a bigger series #4141

- [x] 3. Metadata update semantics
   - [x] The `overrideMetadata` option must be included in the option tables in all operations supporting it (all updates, basically) [ad041ba](https://github.com/telefonicaid/fiware-orion/pull/4176/commits/ad041ba32abcd76e9ff4a0a6287cf00d33c26fa9)
- [x] 10. Legacy attribute format in notifications
   - [x] Maps to "attrsFormat" in the corresponding table in the subscription datamodel https://github.com/telefonicaid/fiware-orion/pull/4176/commits/3dddbbdc99cd0d002a6822ef3f771eaa8dfe92f2
- [x] 17. noAttrDetail option
   - [x] To be included in the corresponding options tables https://github.com/telefonicaid/fiware-orion/pull/4176/commits/a51f18d03ae2b65d85c30943e533aaa7488c6787
- [x] 18. Notification throttling
   - [x] Maps to "throttling" in the corresponding table in the subscription datamodel https://github.com/telefonicaid/fiware-orion/pull/4176/commits/e535c3c5488a548df6fac0baf32bb07cb060996a
   - [x] If the text is too long to be included in one table cell, then consider the option of describing it below the table (the same "trick" used to describe notification rules in subscription.subject.condition section)
- [x] 20. Oneshot subscriptions https://github.com/telefonicaid/fiware-orion/commit/26e978df20257ef6a10c099a0a0d03febfd13bc5
   - [x] Include new status "oneshot" in the corresponding table in the subscription datamodel
- [x] 21. Subscriptions based in alteration type https://github.com/telefonicaid/fiware-orion/pull/4176/commits/d7031c96b1da5e99ef6d3b84e07259bab5764cf3
   - [x] Include new field "alterationTypes" in the corresponding table in the subscription datamodel
- [x] 25. MQTT notifications https://github.com/telefonicaid/fiware-orion/pull/4176/commits/d1c45a04d1e7f3f6cd79fbe68a6c372994d02b6e
   - [x] Include the "mqtt" and "mqttCustom" fields within the subscriptions datamodel
   - [x] The mqtt_notifications.md makes sense as independent document (it will be linked from orion-api.md in the proper points)
- [x] 26. Notify only attributes that change https://github.com/telefonicaid/fiware-orion/pull/4176/commits/93d43e9003dd5587ac162e2ef81b6b33222942c6
   - [x] Include "onlyChangedAttrs" field within the subscriptions datamodel
- [x] 27. Covered subscriptions https://github.com/telefonicaid/fiware-orion/pull/4176/commits/ad3357c2ac1e414e93b9ac9eb94aae85d718a29e
   - [x] Include "covered" field within the subscriptions datamodel
- [x] 28. timeout subscriptions option https://github.com/telefonicaid/fiware-orion/pull/4176/commits/7cde7f3bd843a700414560fe8cd6c5410686005f
   - [x] Include "timeout" field within the subscriptions datamodel 
- [x] 29. lastFailureReason and lastSuccessCode subscriptions fields https://github.com/telefonicaid/fiware-orion/pull/4176/commits/d425391c44032de89ee6cbd114e42e667e066ec0
   - [x] Include both fields within the subscriptions datamodel 
- [x] 30. failsCounter and maxFailsLimit subscriptions fields https://github.com/telefonicaid/fiware-orion/pull/4176/commits/26097b918132ffa9638a503a392ca1c9630158a3
   - [x] Include both fields within the subscriptions datamodel 
   - [x] If the text is too long to be included in one table cell (most probably), then consider the option of describing it below the table (the same "trick" used to describe notification rules in subscription.subject.condition section)
- [x] 32. forcedUpdate option https://github.com/telefonicaid/fiware-orion/commit/ee4027a7fe422d4abb2e4029b901439c265703b2
   - [x] To be included in the corresponding options tables
- [x] 33. flowControl option https://github.com/telefonicaid/fiware-orion/commit/a76bfefad4b66c5ac667fd39c59f20952d4d4136
   - [x] To be included in the corresponding options tables
- [x] 35. skipForwarding option https://github.com/telefonicaid/fiware-orion/commit/6bddb2067d9765b9a284cf6c5cfa451888859531
   - [x] To be included in the corresponding options tables

### CODE REVIEW

**Comment by mapedraza** [comment]
20. Oneshot subscriptions
    - Include new status "oneshot" in the corresponding table in the subscription datamodel
https://github.com/telefonicaid/fiware-orion/pull/4176/commits/26e978df20257ef6a10c099a0a0d03febfd13bc5

Pending to decide if including oneshot_subscription.md into a section insead of the pointer in implementation notes

**Comment by mapedraza** [comment]
> 25. MQTT notifications
>    - [ ] With regards to mqttCustom, the section "Custom Notifcations" should be also modified

This specific point is moved to PR2 since it means to add further information to an existing section.

**Comment by mapedraza** [comment]
Now this PR is ready to review

**Comment by fgalan** [comment]
Some comments:

* Item 3: `overrideMetadata` also makes sense to be mentioned in in POST /v2/op/update

File: doc/manuals/orion-api.md [hunk 1]
```diff
@@ -893,16 +893,49 @@ If `attrsFormat` is `keyValues` then keyValues partial entity representation mod
...
+		},
+		"statusCode": {
+			"code": "200",
+			"reasonPhrase": "OK"
+		}
+	}]
+}
+```
+
+Notifications must include the `Ngsiv2-AttrsFormat` (when using `attrsFormat` = `legacy` is excluded )
```
Comment (fgalan) [inline doc/manuals/orion-api.md:936]:
```suggestion
Notifications must include the `Ngsiv2-AttrsFormat` (expect when `attrsFormat` is `legacy`) 
```

File: doc/manuals/orion-api.md [hunk 2]
```diff
@@ -1848,6 +1935,8 @@ The values that `options` parameter can have for this specific request are:
 |----------|------------------------------------------------------------------------------------------|
 | `count`  | When used, the total number of types is returned in the HTTP header `Fiware-Total-Count` |
 | `values` | When used, the response payload is a JSON array with a list of entity types              |
+| `noAttrDetail` | When used, the request does not provide attribute type details. `types` list associated to each attribute name is set to `[]`. Using this option, Orion solves these queries much faster (up to 30 seconds). |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:1938]:
```suggestion
| `noAttrDetail` | When used, the request does not provide attribute type details. `types` list associated to each attribute name is set to `[]`. Using this option, Orion solves these queries much faster (in some cases saving from 30 seconds to 0.5 seconds). |
```

File: doc/manuals/orion-api.md [hunk 3]
```diff
@@ -1984,8 +2081,13 @@ A subscription is represented by a JSON object with the following fields:
...
+| `status`       |          | string | Either `active` (for active subscriptions) or `inactive` (for inactive subscriptions). If this field is not provided at subscription creation time, new subscriptions are created with the `active` status, which can be changed by clients afterwards. For expired subscriptions, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for subscriptions experiencing problems with notifications, the status is set to `failed`. As soon as the notifications start working again, the status is changed back to `active`. Additionaly, `oneshot` value is available, firing the notification only once whenever the entity is updated after creating the subscription. Once a notification is triggered, the subscription transitions to "status": "inactive". |
+| `throttling`   | ✓        | number | Minimal period of time in seconds which must elapse between two consecutive notifications. Orion implements this is discarding notifications during the throttling guard period. Thus, notifications may be lost if they arrive too close in time. |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2085]:
```suggestion
| `throttling`   | ✓        | number | Minimal period of time in seconds which must elapse between two consecutive notifications. Orion implements this discarding notifications during the throttling guard period. Thus, notifications may be lost if they arrive too close in time. |
```

File: doc/manuals/orion-api.md [hunk 4]
```diff
@@ -2020,28 +2070,61 @@ Based on the `condition` field, the notification triggering rules are as follow:
...
+| `onlyChangedAttrs` | ✓                 | boolean | List of metadata to be included in notification messages. See [Filtering out attributes and metadata](#filtering-out-attributes-and-metadata) section for more detail.            |
+| `timesSent`        | Only on retrieval | number  | Not editable, only present in GET operations. Number of notifications sent due to this subscription.                                                                              |
+| `lastNotification` | Only on retrieval | ISO8601 | Not editable, only present in GET operations. Last notification timestamp in ISO8601 format.                                                                                      |
+| `lastFailure`      | Only on retrieval | ISO8601 | Not editable, only present in GET operations. Last failure timestamp in ISO8601 format. Not present if subscription has never had a problem with notifications.                   |
+| `lastSuccess`      | Only on retrieval | ISO8601 | Not editable, only present in GET operations. Timestamp in ISO8601 format for last successful notification.  Not present if subscription has never had a successful notification. |
```
[the remark this answers is not in the document]
Comment (mapedraza) [reply doc/manuals/orion-api.md]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4176/commits/4c6dd9bc81215b0211286763c8aa4037c38b058b

File: doc/manuals/orion-api.md [hunk 5]
```diff
@@ -2020,28 +2123,66 @@ Based on the `condition` field, the notification triggering rules are as follow:
...
+| `failsCounter`     | Only on retrieval | number  | Not editable, only present in GET operations. The number of consecutive failing notifications associated to the subscription. `failsCounter` is increased by one each time a notification attempt fails and reset to 0 if a notification attempt successes (`failsCounter` is ommitted in this case).|
+| `maxFailsLimit`    | ✓                 | number  | Establishes a maximum allowed number of consecutive fails. If the number of fails overpasses the value of `maxFailsLimit` (i.e. at a given moment `failsCounter` is greater than `maxFailsLimit`) then Orion automatically passes the subscription to `inactive` state. A subscripiton update operation (`PATCH /v2/subscription/subId`) is needed to re-enable the subscription (setting its state `active` again). |
+
+Regarding `covered` field, as an example, if `attrs` is `[A, B, C]` for a given subscription, the default behavior
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2143]:
```suggestion
Regarding `onlyChangedAttrs` field, as an example, if `attrs` is `[A, B, C]` for a given subscription, the default behavior 
```

File: doc/manuals/orion-api.md [hunk 6]
```diff
@@ -2050,9 +2191,27 @@ An `httpCustom` object contains the following subfields.
 | `qs`      | ✓        | object | A key-map of URL query parameters that are included in notification messages.                 |
 | `method`  | ✓        | string | The method to use when sending the notification (default is POST). Only valid HTTP methods are allowed. On specifying an invalid HTTP method, a 400 Bad Request error is returned.|
 | `payload` | ✓        | string | The payload to be used in notifications. If omitted, the default payload (see [Notification Messages](#notification-messages) sections) is used.|
+| `timeout` | ✓        | number | Maximum time the subscription waits for the response when using in milliseconds. The maximum value allowed for this parameter is 1800000 (30 minutes). If `timeout` is defined to 0 or omitted, then the value passed as `-httpTimeout` CLI parameter is used. See section in the
+[Command line options](admin/cli.md#command-line-options) for more details. |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2195]:
```suggestion
| `timeout` | ✓        | number | Maximum time (in milliseconds) the subscription waits for the response. The maximum value allowed for this parameter is 1800000 (30 minutes). If `timeout` is defined to 0 or omitted, then the value passed as `-httpTimeout` CLI parameter is used. See section in the [Command line options](admin/cli.md#command-line-options) for more details. |
```

File: doc/manuals/orion-api.md [hunk 7]
```diff
@@ -1848,6 +1937,8 @@ The values that `options` parameter can have for this specific request are:
 |----------|------------------------------------------------------------------------------------------|
 | `count`  | When used, the total number of types is returned in the HTTP header `Fiware-Total-Count` |
 | `values` | When used, the response payload is a JSON array with a list of entity types              |
+| `noAttrDetail` | When used, the request does not provide attribute type details. `types` list associated to each attribute name is set to `[]`. Using this option, Orion solves these queries much faster (in some cases saving from 30 seconds to 0.5 seconds). |
+| `skipForwarding` | when used, CB skips forwarding to CPrs. The query is evaluated using exclusively CB local context information. |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:1941]: Doesn't apply to GET /v2/types

```suggestion
```

File: doc/manuals/orion-api.md [hunk 8]
```diff
@@ -1922,16 +2013,24 @@ This operation returns a JSON object with information about the type:
...
+| Parameter    | Optional | Type   | Description         | Example |
+|--------------|----------|--------|---------------------|---------|
+| `entityType` |          | string | Entity Type.        | `Room`  |
+| `options`    | ✓        | string | Options dictionary. | `count` |
+
+The values that `options` parameter can have for this specific request are:
+
+| Options  | Description                                                                              |
+|----------|------------------------------------------------------------------------------------------|
+| `skipForwarding` | when used, CB skips forwarding to CPrs. The query is evaluated using exclusively CB local context information. |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2025]: Doesn't apply to GET /v2/types

```suggestion
```

File: doc/manuals/orion-api.md [hunk 9]
```diff
@@ -1922,16 +2012,17 @@ This operation returns a JSON object with information about the type:
...
+|--------------|----------|--------|---------------------|---------|
+| `entityType` |          | string | Entity Type.        | `Room`  |
 
 _**Request headers**_
 
 | Header               | Optional | Description                                                                                    | Example            |
 |----------------------|----------|------------------------------------------------------------------------------------------------|--------------------|
 | `Fiware-Service`     | ✓        | Tenant or service. See subsection [Multitency](#multitenancy) for more information.            | `acme`             |
 | `Fiware-ServicePath` | ✓        | Service path or subservice. See subsection [Service Path](#service-path) for more information. | `/project`         |
+| `noAttrDetail` | When used, the request does not provide attribute type details. `types` list associated to each attribute name is set to `[]`. Using this option, Orion solves these queries much faster (in some cases saving from 30 seconds to 0.5 seconds). |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2025]: `noAttrDetail`, is an `option`, not a header.

```suggestion
```

File: doc/manuals/orion-api.md [hunk 10]
```diff
@@ -1922,9 +2012,15 @@ This operation returns a JSON object with information about the type:
 
 _**Request query parameters**_
 
-| Parameter    | Optional | Type   | Description  | Example |
-|--------------|----------|--------|--------------|---------|
-| `entityType` |          | string | Entity Type. | `Room`  |
+| Parameter    | Optional | Type   | Description         | Example |
+|--------------|----------|--------|---------------------|---------|
+| `entityType` |          | string | Entity Type.        | `Room`  |
+
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2018]: Note that entityType is not actually a query parameter, but an URL parameter, e.g.: `GET /v2/types/T1?options=noAttrDetail`)

```suggestion
| `options` | ✓        | string | Options dictionary.                        | `count` |
```

File: doc/manuals/orion-api.md [hunk 11]
```diff
@@ -1922,9 +2012,14 @@ This operation returns a JSON object with information about the type:
 
 _**Request query parameters**_
 
-| Parameter    | Optional | Type   | Description  | Example |
-|--------------|----------|--------|--------------|---------|
-| `entityType` |          | string | Entity Type. | `Room`  |
+| Parameter    | Optional | Type   | Description         | Example |
+|--------------|----------|--------|---------------------|---------|
+| `options` | ✓        | string | Options dictionary.                        | `count` |
+The values that `options` parameter can have for this specific request are:
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2018]:
```suggestion
| `options` | ✓        | string | Options dictionary.                        | `noAttrDetail` |

The values that `options` parameter can have for this specific request are:
```

**Pull request merged**